### DISCUSSION TITLE
KeyValue Revision not unique to a key but to a bucket: Difference to a spec?

### DISCUSSION THREAD

**Discussion by mikoskinen** [nats-io/nats-server#4580] [General]
I've been using the KeyValue store of NATS quite heavily lately and really like it. But the revision management/revision history of a particular key is kind of a strange. This has been previously discussed here: https://github.com/nats-io/nats-server/discussions/3466

Long story short, I initially believed that the .NET Client of NATS was doing something incorrectly after noticing that the revisions for a specific keys weren't 1, 2, 3, 4, 5 etc. as they instead are affected by the other keys in the bucket.

I checked the spec (from https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-8.md) and it mentions that:

// Revision is a unique sequence for this value

I decided to verify the bug using NATS Cli and as it was behaving like the .NET client, I ended up in the discussion mentioned above which shows that things are working as expected.

Couple questions/clarifications:
1. The spec mentions that "Revision is a unique sequence for this value" but I'm not sure if this is correct given that the sequence is shared between the keys in the bucket.
2. Is there anything in the roadmap which would change the current implementation or which could add a new feature to allow a unique key specific history?

Thanks in advance and thank you for the great project.

**Comment by ripienaar** [comment]
Values are actually jetstream messages and the sequence is the message sequence of the stream - since the underlying stream is sequenced by message and not subject (every message in the stream needs a unique sequence) this is also how it behaves in KV.

Not sure there is anything we can do about this in the short term - but ultimately this is working as designed.

**Comment by jnmoyne** [comment]
> The spec mentions that "Revision is a unique sequence for this value" but I'm not sure if this is correct given that the sequence is shared between the keys in the bucket.

And indeed every sequence number for subsequent values is is unique per key, just not contiguous.

If you do a watch for the key in question you do get the values in order so you can number them as you receive them.
